Add environment config file existence check

daniel-niknam opened the pull request:
### Motivation / Background

Fixes #55779

### Detail

This Pull Request adds a behaviour to `Rails::Application` which will raise an error if unable to load any environment file.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil commented:
> because I think this would warn the user every time a Rails::Engine is loaded

Right, I don't think this is what we want. You should probably put the initializer in `Rails::Application` itself (?), which subclasses `Rails::Engine`, so that it only warns when the top level files are missing.

byroot commented:
Ok, I pushed a commit to show how I think it would work. A bunch of tests would need cleanup though.

byroot commented:
So aside from the vast majority of Railties tests needing to be updated, another potential blocker is:

```ruby
    test "the application can be eager loaded even when there are no frameworks" do
      FileUtils.rm_rf("#{app_path}/app/jobs/application_job.rb")
      FileUtils.rm_rf("#{app_path}/app/models/application_record.rb")
      FileUtils.rm_rf("#{app_path}/app/mailers/application_mailer.rb")
      FileUtils.rm_rf("#{app_path}/config/environments")
      add_to_config <<-RUBY
        config.enable_reloading = false
        config.eager_load = true
      RUBY

      use_frameworks []

      assert_nothing_raised do
        app "development"
      end
    end
```

Rails is meant to be bootable from single file. So it's legit not to have the corresponding env file.

A solution might be to make the logic smarter, e.g.:

```ruby
if env_file_exists?
  require it
elsif any_env_file_exists?
  raise error
else
  # don't care.
end
```

daniel-niknam commented:
> I'm actually surprised `:load_environment_config` is executed for every engine. Granted I'm not the most familoar with initializers.
> 
> Is it expected that engines have their own `config/environments/*.rb` files?

@byroot Yes, we are on the same boat 😄 I suspect that would be the case because both `Rails::Application` and any engines will be using the same underlying class, which is the `Rails::Engine`.

I confirmed this behaviour by generating a Rails app first, generating a plugin inside the Rails app, creating an environment file in `config/environments/development.rb` inside the plugin with a content of `raise "inside engine"`, and after running Rails, I got the error `/private/tmp/blah/blorgh/config/environments/development.rb:1:in '<main>': inside engine (RuntimeError)`. 

So yes, the new engine is not generated with a set of environment files, but if you create it, it's going to load them. I concluded that means the block of initialiser is executed in any engine instance.

daniel-niknam commented:
> > because I think this would warn the user every time a Rails::Engine is loaded
> 
> Right, I don't think this is what we want. You should probably put the initializer in `Rails::Application` itself (?), which subclasses `Rails::Engine`, so that it only warns when the top-level files are missing.

Thanks for the feedback @skipkayhil. I didn't implement it inside the `Rails::Application` class because I couldn't see any existing pattern that looks the same. But I think I need to find a way to make sure that logic is only executed on the main application, so I might end up introducing it. I want to play with that `group:` attribute to see if that does something useful for this use case 😄

daniel-niknam commented on railties/test/commands/credentials_test.rb:
```diff
@@ -129,6 +129,7 @@ class Rails::Command::CredentialsTest < ActiveSupport::TestCase
   end
 
   test "edit command does not raise when an initializer tries to access non-existent credentials" do
+    File.write(app_path("config", "environments", "qa.rb"), "")
```
In this test, we already have `development.rb`, `test.rb` and `production.rb`. I've added the extra env file only in the test. The other way would be to extend the `setup :build_app` somehow to accept extra env. Maybe out of scope of this PR.

daniel-niknam reviewed (commented):
I managed to have another look at this PR finally 😅 We’ve had a bit of a cold going around the house, everyone caught it one by one, and I was the last one standing. Finally caught it myself, but all good now! 😄 

@byroot When you can, can you please have a look at this PR one more time? Before I do a cleanup, it would be good to know what you think. I added some comments about some possible changes, but I'm not sure because I'm new.

rafaelfranca commented on railties/test/application/loading_test.rb:
```diff
@@ -111,6 +111,10 @@ class User < ActiveRecord::Base
   end
 
+  test "Raises RuntimeError when config/environments/ENV.rb file does not exist" do
+    assert_raise(RuntimeError, /Rails environment has been set to non_existent_environment but config\/environments\/non_existent_environment.rb does not exist./) { boot_app "non_existent_environment" }
```
```suggestion
    assert_raises(RuntimeError, match: /Rails environment has been set to non_existent_environment but config\/environments\/non_existent_environment.rb does not exist./) { boot_app "non_existent_environment" }
```

byroot commented on railties/lib/rails/application.rb:
```diff
@@ -636,6 +636,13 @@ def ensure_generator_templates_added
 
     private
+      def missing_environment_file
+        raise <<~TEXT
```
Nah, it's extra complexity for nothing. They can create an empty file if needed.

byroot commented:
Generally looks good to me, once Rafael feedback is addressed and the git history cleaned up I think it's ready for merge. Will probably wait for 8.1 final to be released though.

daniel-niknam commented on railties/lib/rails/application.rb:
```diff
@@ -636,6 +636,13 @@ def ensure_generator_templates_added
+          Rails environment has been set to #{Rails.env} but
+          config\/environments\/#{Rails.env}.rb does not exist"
+        TEXT
+      end
```
I leave it out for now, but I managed to get it working with something like:

```ruby
      def missing_environment_file
        corrections = DidYouMean::SpellChecker.new(dictionary: all_environment_files.map {|path| path.basename(".rb") }).correct(Rails.env)

        message = "Rails environment has been set to #{Rails.env} but config/environments/#{Rails.env}.rb does not exist."
        message += DidYouMean::Formatter.message_for(corrections) if corrections.any?

        raise message
      end
```

daniel-niknam committed:
Add environment config file existence check

This commit fixes the following reported issue:

Fix #55779

Summary

When running Rails applications with a non-existent environment, Rails will silently continue without a proper error message.

Example:

```bash
RAILS_ENV=NON_EXISTING_ENV bin/rails server ...
```

This commit will add the following functionality:

- Enhances `load_environment_config` initializer to raise an error when there is one or many environment files in `config/environments` or any other configured location _but_ rails is unable to load any environment file in any configured location.
- Adds two methods to Rails::Engine
  - private #any_environment_files? which returns an empty array by default to prevent Rails::Engine, aka plugins, from enforcing this functionality
  - private #missing_environment_file which returns nil by default to prevent Rails::Engine aka plugins to enforce this functionality
- Adds two methods to Rails::Application
  - private #any_environment_files? which returns true if any environment file is available to be loaded in any configured location. The reason for checking existing environment files is to prevent raising an error if no environment files are available. This is useful in cases like running bug [report templates](https://github.com/rails/rails/blob/main/guides/bug_report_templates/generic.rb).
  - private #missing_environment_file, which raises a RuntimeError to let the user know that Rails is unable to locate the environment file.

Co-authored-by: Rafael Mendonça França <rafael@rubyonrails.org>
Co-authored-by: Jean Boussier <jean.boussier@gmail.com>
Co-authored-by: zzak <zzak@hey.com>

daniel-niknam commented:
I just realised I messed up the git history again 🤦 😅 It's ready now. Thank you 🙇

byroot commented on railties/lib/rails/engine.rb:
```diff
@@ -687,6 +693,12 @@ def run_tasks_blocks(*) # :nodoc:
+
+      def any_environment_files?
+        []
+      end
```
```suggestion
      def any_environment_files?
        false
      end
```

daniel-niknam replied:
Dang it. I must have missed it during rebase 😂
